pkg/cache/cluster.go

```diff
@@ -596,23 +615,32 @@ func (c *clusterCache) sync() error {
 
 // EnsureSynced checks cache state and synchronizes it if necessary
 func (c *clusterCache) EnsureSynced() error {
-	// first check if cluster is synced *without lock*
+	syncStatus := c.syncStatus
+
+	// first check if cluster is synced *without acquiring the full clusterCache lock*
+	syncStatus.lock.Lock()
 	if c.synced() {
-		return c.syncError
+		syncError := syncStatus.syncError
+		syncStatus.lock.Unlock()
+		return syncError
 	}
+	syncStatus.lock.Unlock() // release the lock, so that we can acquire the parent lock (see struct comment re: lock acquisition ordering)
 
 	c.lock.Lock()
+	syncStatus.lock.Lock()
 	defer c.lock.Unlock()
+	defer syncStatus.lock.Unlock()
+
 	// before doing any work, check once again now that we have the lock, to see if it got
 	// synced between the first check and now
 	if c.synced() {
-		return c.syncError
+		return syncStatus.syncError
 	}
 	err := c.sync()
```

Comment:
So we are running this expensive method while holding the new lock here. What is the point of the new lock then if it's held here? Just trying to understand.

Reply:
A good question.... my main goal here was to preserve the behaviour of the existing code, while making it data race safe.

The situation the original code was (AFAICT) trying to avoid is forcing external calls to `EnsureSynced()` (which ensures that the cluster cache is synced to within the given timeframe) to block on expensive operations (eg cluster I/O) that might be in progress at the same time.

The `clusterCache.lock` lock is generally reserved for code that will (necessarily) take a while, due to the need to wait for cluster I/O:
- refreshing the cluster cache
- starting/cancelling watches
- calling resourceUpdate/event handlers
- calling other k8s cluster APIs

In contrast, `clusterCacheSync.lock` is generally reserved for code that desires fast access:
- The first half ("happy path") of `EnsureSynced()`
    - comparing `syncTime` with `resyncTimeout` (`synced()`)
- setting and retrieving the field values 

The _only_ case where `clusterCacheSync.lock` is held on a slow operation:
- the second half of `EnsureSynced()` (the cache invalid, non "happy path" part of `EnsureSynced()`)
    -  And this is consistent with the expected behaviour of `EnsureSynced()` by callers... if the cluster is not synced, we do not want to return from this method until it _is_ synced.

So ideally, in the vast majority of cases, calls to `EnsureSynced()` by consuming code will be able to complete by only acquiring the fast lock (`clusterCacheSync.lock`), and thus the calling code will not need to block on the slower lock.

In the minority of cases (when the cache has not yet synced or the cache sync has expired) the full lock will need to be acquired.

Reply:
I see. I wonder if we can just use read lock of `lock` to protect reads. That would simplify things quite a bit, but I'm not clear whether it would or would not create any problems. It seems to me that logically `synched` means all the state is up to date. So I'm not sure how this check can be performed out of bound, not holding the main lock. To put it differently, how can we know if it's synchronized if a mutative operation is happening right now? We need to wait for it to finish, don't we?

@alexmt Could you clarify how this should work please?

Reply:
I would argue that the alternative, using a read lock, is unfortunately more complex. 

Why? With an RWLock, write access on that lock will block all the reads, so in order to judge the cost  of using a read lock to fix this data race, we first need to look at who acquires the write access.

So the main acquirer here is `runSynced()` (in the same file), which is used to run a bunch of different functions... Unfortunately, the contained functions contain indirect calls to code that is not only outside this file, but outside of the package, and even outside of gitops-engine entirely. So the scope and potential cost of write lock ownership is broad.

Doing some code tracing: in a nearly all of the cases where the write lock is acquired by `runSynced()`, `onNodeRemoved` and `onNodeUpdated` are called. These two functions will then call the `resourceUpdateHandlers` for the `clusterCache`.

The `resourceUpdateHandlers` are functions which are injected by `OnResourceUpdated`, and `OnResourceUpdated` is called by consuming code that is both inside and outside gitops engine. 
- In gitops engine: It appears that a call to `syncContext.Sync()` will block in a resourceUpdateHandler, so this alone makes the write lock very heavy.
- Outside gitops engine: `liveStateCache.onObjectUpdated` may involve cluster I/O, and we would probably want to avoid blocking a lock on network I/O where possible.


In any case, in order to judge how heavy this lock is, you have to do a LOT of code tracing across packages/components, and (extra unfortunately) you would need to make sure that the desired properties (not holding the write lock for too long) of these packages/components never change in a way that would make the write lock (in this PR) more expensive.

So IMHO I don't think it's feasible to rely on calling code not to break us in this case, and that unfortunately means the introduction of the second lock, as done in the PR.

Though, in contrast to the above described complexity:
- This PR sets a strict set of locking rules that only apply to a single field in a struct, and only apply within a single file.
- This PR preserves the existing concurrency properties of the code (eg the code should work exactly the same as before, neither deadlocking or blocking in unexpected ways)

Reply:
@jgwest Thanks for a very detailed response! 👍 I think it makes sense, thanks for investigating the issue so thoroughly.

Reply:
Ops, missed this PR. Thank you @jgwest ! Looking